Add `title` field to base `Card` type

lukemelia opened the pull request:
- Introduced `CardBase` base class that `Card` and primitive fields extends from
- Added `title` field to Card of type StringCard
- Updated test and demo cards to define `title` as a computed StringCard where appropriate

lukemelia committed:
Add `title` field to base `Card` type

- Introduced `Primitive` base class that `Card` and primitive fields extends from
- Added `title` string field to Card
- Updated test and demo cards to define `title` as a computed string where appropriate

lukemelia commented:
I'm eager for feedback on this. I have two things that I am leaning toward changing if people agree:

1) Change the base class name from `Primitive` to something else, like `CardPrimitive` or `CardBase`.
2) Add the field `cardTitle` instead of `title` in order to avoid conflicts with business domains like Blog Post, Booking, and Employee.

habdelra commented:
i vaguely recall in the previous versions of our API that we had special namespace for "system" fields
i believe it was an `_` character, such that, any field that began with an `_` was a system field, and all other fields were userland fields. And the base card, i believe, had `_id`, `_title` and `_description` fields.

habdelra commented:
Also, I'm open to `CardBase` or `CardPrimitive` for the name of the root card class.

lukemelia committed:
WIP spec fixes

jurgenwerk commented:
I like `CardBase`, it sounds more descriptive

lukemelia committed:
Refactor to better naming

Rename Primitive to CardBase
Make _metadata a computed field contains(MetadataCard)

lukemelia commented:
Something to look for while you review: are there functions that should be typed to accept `Card` instead of `CardBase`? When I refactored, I converted almost all usages of the type `Card` to `CardBase`, but perhaps some places should require `Card`?

habdelra commented on packages/base/string.gts:
Perhaps we might want to reexport the StringCard from the api to stay consistent with the other fields?

habdelra replied:
Lol, never mind, I see you do that already

habdelra reviewed (approved):
Looks good!

habdelra commented on packages/base/card-ref.gts:
This should probably use Card type since it’s a composite card not a primitive card

habdelra commented on packages/base/catalog-entry.gts:
This should probably use Card type since it’s a composite card not a primitive card

habdelra commented on packages/base/card-api.gts:
```diff
@@ -129,9 +130,9 @@ function isStaleValue(value: any): value is StaleValue {
     return false;
   }
 }
-const deserializedData = new WeakMap<Card, Map<string, any>>();
```
This should probably use Card type since it’s referring to a map of fields that only composite cards contain

habdelra commented on packages/base/card-api.gts:
```diff
@@ -129,9 +130,9 @@ function isStaleValue(value: any): value is StaleValue {
-const identityContexts = new WeakMap<Card, IdentityContext>();
+const deserializedData = new WeakMap<CardBase, Map<string, any>>();
+const recomputePromises = new WeakMap<CardBase, Promise<any>>();
+const identityContexts = new WeakMap<CardBase, IdentityContext>();
```
This is probably a Card type since only composite cards can have an identity

habdelra commented on packages/base/card-api.gts:
```diff
@@ -173,7 +174,7 @@ export async function flushLogs() {
 
 export class IdentityContext {
-  readonly identities = new Map<string, Card>();
+  readonly identities = new Map<string, CardBase>();
```
This is probably a Card type since only composite cards can have an identity

habdelra commented on packages/base/card-api.gts:
```diff
@@ -682,7 +684,7 @@ class LinksTo<CardT extends CardConstructor> implements Field<CardT> {
   }
 
-  getter(instance: Card): CardInstanceType<CardT> {
+  getter(instance: CardBase): CardInstanceType<CardT> {
```
linksTo cards are never primitive, perhaps this is a Card type.

habdelra commented on packages/base/card-api.gts:
```diff
@@ -693,7 +695,7 @@ class LinksTo<CardT extends CardConstructor> implements Field<CardT> {
   }
 
-  queryableValue(instance: any, stack: Card[]): any {
+  queryableValue(instance: any, stack: CardBase[]): any {
```
linksTo cards are never primitive, perhaps this is a Card type.

habdelra commented on packages/base/card-api.gts:
```diff
@@ -790,7 +792,7 @@ class LinksTo<CardT extends CardConstructor> implements Field<CardT> {
     _fieldMeta: undefined,
     identityContext: IdentityContext,
-    _instancePromise: Promise<Card>,
+    _instancePromise: Promise<CardBase>,
```
linksTo cards are never primitive, perhaps this is a Card type.

habdelra commented on packages/base/card-api.gts:
```diff
@@ -819,11 +821,11 @@ class LinksTo<CardT extends CardConstructor> implements Field<CardT> {
   }
 
-  validate(_instance: Card, value: any) {
+  validate(_instance: CardBase, value: any) {
```
linksTo cards are never primitive, perhaps this is a Card type.

habdelra commented on packages/base/card-api.gts:
```diff
@@ -844,7 +846,7 @@ class LinksTo<CardT extends CardConstructor> implements Field<CardT> {
   }
 
-  async handleNotLoadedError<T extends Card>(
+  async handleNotLoadedError<T extends CardBase>(
```
linksTo cards are never primitive, perhaps this is a Card type.

lukemelia replied:
I think I can't change this due to inheritance

habdelra commented:
I started marking it up, but perhaps it’s just easier to make the general comment that LinksTo and LinksToMany fields have cards that are never primitive, these are always composite cards. So in the Field classes for these we can probably use the Card type and not the CardBase type as the parameterized type. I think, tho, it’s fine to merge what you have and do a refactor as a follow on for more precise types.

lukemelia committed:
Re-type LinksTo/LinkToMany to Card instead of CardBase

lukemelia committed:
Add title field to card instead of _metadata

lukemelia committed:
Typing progress

lukemelia committed:
Update tests

ef4 commented on packages/base/card-api.gts:
```diff
@@ -1465,9 +1470,33 @@ class IDCard extends Card {
+export class Card extends CardBase {
+  @field title = contains(StringCard, {
+    computeVia: function (this: Card) {
+      return 'Untitled';
```
We discussed possibly making this a normal field, not a computed, since it's already legal to override that with a computed anyway when you want that.

I think that's probably still a good idea. It doesn't change the case for child cards that want to override with a computed, and it means that other ones that want a manual title would have that (which they cannot otherwise do unless we add a storage API for computeds).

lukemelia replied:
I will make the change, but I'll note that I discovered it is currently also "legal" to override a computed with a non-computed.

lukemelia committed:
Remove default Untitled value of Card#title field
